mqm_tool.py

```diff
@@ -0,0 +1,391 @@
+        if stop_condition(count_zero_list, count_list, grid_percent, count_num, cell_num, out_distribution):
+            # wirte out one row
+            summary_table.append(util.summary_table_row_generation(entire_data, name_num, round(initial_area * 1e-6, 2), grid_area, folder_name))
+
+            # write out a Geojson file
+            util.geojson_write(depth_count, bb_collec, os.path.join(folder_path, geojson_path), cell_num, grid_area,
+                               None, 'tree_v1', hist, flag_val = True)
+            del util
+            
+            # perform the 2nd k-d tree
+            if max_count != -1:                
+                new_grids_list = []
+                new_counts_list = []
+                new_area_list = []
+                new_grid_id_list = []
+                grid_ids = []
+                gridid_start = 1
+                
+                # find all grid indices in which the count is greater the max count
+                big_grid_index_list = [index_value for index_value in range(hist.shape[0]) if hist[index_value] > max_count]
+                if big_grid_index_list:
+                    # refine big grids through applying the 2nd K-D tree
+                    for extension_ind in big_grid_index_list:
+                        for depth_num in range(1, int(maximum_level) + 1):
+                            new_bb_collec, new_counts_collec, new_grid_id_collec = extend_partition(depth_num, bb_collec[extension_ind], entire_data, gridid_start)
+                            new_grid_id_list.append(new_grid_id_collec)
+
+                            # calculate areas
+                            new_area = grid_area / (2** depth_num)
+                            new_area_list.append(new_area)
+
+                            # update the start point of the grid id
+                            gridid_start += len(new_bb_collec)
+
+                            # stop condition
+                            if len([x for x in new_counts_collec if x < max_count]) == len(new_counts_collec):
+                                for small_ind in range(len(new_bb_collec)):
+                                    grid_ids.append(gridid_start + small_ind - len(new_bb_collec))
+                                    new_grids_list.append(new_bb_collec[small_ind])
+                                    new_counts_list.append(new_counts_collec[small_ind])
+                                break
+                    
+                    util = Utility()
+                    # write out a Geojson file
+                    util.geojson_write(depth_count, new_grids_list,
+                                       os.path.join(folder_path, geojson_path), None, None, grid_ids, 'cascade-kdtree', new_counts_collec, flag_val = True)
+                    del util
+            
+            # road counts
+            if road_file:
+                road_count(road_file, bb_collec, hist, folder_path, out_BB)
```

Comment:
I'm afraid I still don't understand the importance of this. It appears that you are now writing out the flag counts per grid of the first geojson file in a folder. Why are the flag counts from this file so important? Also, it looks like you are rewriting the same file for every depth count.

Reply:
Why are the flag counts from this file so important?
The reason why the flag counts are important is that these flag counts can be used to statistical comparisons between error counts and the flag counts. Specifically, road_file is a Geojson file with all the road coordinates. Given the Geojson file, the road_count function is going to compute flag counts of the geometries per grid through using all the grids that are generated by the first k-d tree.

Also, it looks like you are rewriting the same file for every depth count.
Actually, no, the program only performs road_count once the stop condition is achieved, which means the function will be executed for the last depth count.

Reply:
Ok, that clears up a lot. Thanks for the explanation. I am still curious about why this is the first geojson in the first sub-directory, and how this file is generated. Are you expecting that atlas-checks will generate this file? or am I missing where mqm is generating this file?

Reply:
The program was not necessary to find the first geojson in the first sub-directory. In my case, I only have the road file for UWA, and other directories (e.g. CYP) do not have associated road files. The program thus only runs the function when processing UWA, but the program, in reality, would be able to automatically execute the function if the directory consisted of the road file inside a sub-directory.

Currently, the road file is similar to other geojson files, which means users can determine whether or not they want the file. I will figure out how this file is generated.